File: CS1/AlgorithmExamples.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace CS1
{
    public class AlgorithmExamples
    {
        // 面试题1: 两数之和 - 给定一个数组和一个目标值，找出数组中和为目标值的两个数的索引
        public static int[] TwoSum(int[] nums, int target)
        {
            // 使用字典存储数字和索引
            Dictionary<int, int> numDict = new Dictionary<int, int>();

            for (int i = 0; i < nums.Length; i++)
            {
                int complement = target - nums[i];

                if (numDict.ContainsKey(complement))
                {
                    return new int[] { numDict[complement], i };
                }

                if (!numDict.ContainsKey(nums[i]))
                {
                    numDict.Add(nums[i], i);
                }
            }

            return new int[0]; // 没找到解决方案
        }

        // 面试题2: 回文字符串检查
        public static bool IsPalindrome(string s)
        {
            if (string.IsNullOrEmpty(s))
                return true;

            // 预处理：只保留字母和数字，并转换为小写
            string processed = new string(s.Where(c => char.IsLetterOrDigit(c)).ToArray()).ToLower();

            // 使用双指针法
            int left = 0;
            int right = processed.Length - 1;

            while (left < right)
            {
                if (processed[left] != processed[right])
                    return false;

                left++;
                right--;
            }

            return true;
        }

        // 面试题3: 找出数组中的最大子数组和 (Kadane算法)
        public static int MaxSubArray(int[] nums)
        {
            if (nums == null || nums.Length == 0)
                return 0;

            int currentMax = nums[0];
            int globalMax = nums[0];

            for (int i = 1; i < nums.Length; i++)
            {
                // 当前最大子数组和：取当前元素或者当前元素加上之前的最大子数组和
                currentMax = Math.Max(nums[i], currentMax + nums[i]);

                // 更新全局最大值
                globalMax = Math.Max(globalMax, currentMax);
            }

            return globalMax;
        }

        // 面试题4: 合并两个有序数组
        public static void MergeSortedArrays(int[] nums1, int m, int[] nums2, int n)
        {
            int i = m - 1;  // nums1的末尾位置
            int j = n - 1;  // nums2的末尾位置
            int k = m + n - 1;  // 合并后数组的末尾位置

            // 从后向前填充nums1
            while (i >= 0 && j >= 0)
            {
                if (nums1[i] > nums2[j])
                {
                    nums1[k] = nums1[i];
                    i--;
                }
                else
                {
                    nums1[k] = nums2[j];
                    j--;
                }
                k--;
            }

            // 如果nums2还有剩余元素，全部复制到nums1
            while (j >= 0)
            {
                nums1[k] = nums2[j];
                j--;
                k--;
            }
        }

        // 面试题5: 反转链表
        public static ListNode ReverseLinkedList(ListNode head)
        {
            ListNode prev = null;
            ListNode current = head;

            while (current != null)
            {
                ListNode nextTemp = current.Next;  // 保存下一个节点
                current.Next = prev;  // 反转指针
                prev = current;  // 移动prev指针
                current = nextTemp;  // 移动current指针
            }

            return prev;  // 新的头结点
        }

        // 面试题6: 使用栈实现队列
        public class MyQueue
        {
            private Stack<int> inStack;  // 用于入队操作
            private Stack<int> outStack; // 用于出队操作

            public MyQueue()
            {
                inStack = new Stack<int>();
                outStack = new Stack<int>();
            }

            // 入队操作
            public void Push(int x)
            {
                inStack.Push(x);
            }

            // 出队操作
            public int Pop()
            {
                if (outStack.Count == 0)
                {
                    // 如果outStack为空，将inStack中的所有元素倒入outStack
                    while (inStack.Count > 0)
                    {
                        outStack.Push(inStack.Pop());
                    }
                }

                return outStack.Pop();
            }

            // 查看队首元素
            public int Peek()
            {
                if (outStack.Count == 0)
                {
                    // 如果outStack为空，将inStack中的所有元素倒入outStack
                    while (inStack.Count > 0)
                    {
                        outStack.Push(inStack.Pop());
                    }
                }

                return outStack.Peek();
            }

            // 检查队列是否为空
            public bool Empty()
            {
                return inStack.Count == 0 && outStack.Count == 0;
            }
        }

        // 面试题7: 二叉树的最大深度 (递归解法)
        public static int MaxDepth(TreeNode root)
        {
            if (root == null)
                return 0;

            int leftDepth = MaxDepth(root.Left);
            int rightDepth = MaxDepth(root.Right);

            // 当前节点的最大深度 = max(左子树深度, 右子树深度) + 1
            return Math.Max(leftDepth, rightDepth) + 1;
        }

        // 面试题8: 有效的括号
        public static bool IsValidParentheses(string s)
        {
            if (string.IsNullOrEmpty(s))
                return true;

            Stack<char> stack = new Stack<char>();

            foreach (char c in s)
            {
                if (c == '(' || c == '[' || c == '{')
                {
                    stack.Push(c);
                }
                else
                {
                    if (stack.Count == 0)
                        return false;

                    char top = stack.Pop();

                    if (c == ')' && top != '(')
                        return false;
                    if (c == ']' && top != '[')
                        return false;
                    if (c == '}' && top != '{')
                        return false;
                }
            }

            return stack.Count == 0;
        }
    }

    // 支持类：链表节点
    public class ListNode
    {
        public int Val;
        public ListNode Next;

        public ListNode(int val = 0, ListNode next = null)
        {
            Val = val;
            Next = next;
        }
    }

    // 支持类：二叉树节点
    public class TreeNode
    {
        public int Val;
        public TreeNode Left;
        public TreeNode Right;

        public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
        {
            Val = val;
            Left = left;
            Right = right;
        }
    }
}

File: Visual Basic/AlgorithmExamples.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Namespace VBExamples
    Public Class AlgorithmExamples
        ' 面试题1: 两数之和 - 给定一个数组和一个目标值，找出数组中和为目标值的两个数的索引
        Public Shared Function TwoSum(nums() As Integer, target As Integer) As Integer()
            ' 使用字典存储数字和索引
            Dim numDict As New Dictionary(Of Integer, Integer)()
            
            For i As Integer = 0 To nums.Length - 1
                Dim complement As Integer = target - nums(i)
                
                If numDict.ContainsKey(complement) Then
                    Return New Integer() {numDict(complement), i}
                End If
                
                If Not numDict.ContainsKey(nums(i)) Then
                    numDict.Add(nums(i), i)
                End If
            Next
            
            Return New Integer() {} ' 没找到解决方案
        End Function
        
        ' 面试题2: 回文字符串检查
        Public Shared Function IsPalindrome(s As String) As Boolean
            If String.IsNullOrEmpty(s) Then
                Return True
            End If
            
            ' 预处理：只保留字母和数字，并转换为小写
            Dim processed As String = New String(s.Where(Function(c) Char.IsLetterOrDigit(c)).ToArray()).ToLower()
            
            ' 使用双指针法
            Dim left As Integer = 0
            Dim right As Integer = processed.Length - 1
            
            While left < right
                If processed(left) <> processed(right) Then
                    Return False
                End If
                
                left += 1
                right -= 1
            End While
            
            Return True
        End Function
        
        ' 面试题3: 找出数组中的最大子数组和 (Kadane算法)
        Public Shared Function MaxSubArray(nums() As Integer) As Integer
            If nums Is Nothing OrElse nums.Length = 0 Then
                Return 0
            End If
            
            Dim currentMax As Integer = nums(0)
            Dim globalMax As Integer = nums(0)
            
            For i As Integer = 1 To nums.Length - 1
                ' 当前最大子数组和：取当前元素或者当前元素加上之前的最大子数组和
                currentMax = Math.Max(nums(i), currentMax + nums(i))
                
                ' 更新全局最大值
                globalMax = Math.Max(globalMax, currentMax)
            Next
            
            Return globalMax
        End Function
        
        ' 面试题4: 合并两个有序数组
        Public Shared Sub MergeSortedArrays(nums1() As Integer, m As Integer, nums2() As Integer, n As Integer)
            Dim i As Integer = m - 1 ' nums1的末尾位置
            Dim j As Integer = n - 1 ' nums2的末尾位置
            Dim k As Integer = m + n - 1 ' 合并后数组的末尾位置
            
            ' 从后向前填充nums1
            While i >= 0 AndAlso j >= 0
                If nums1(i) > nums2(j) Then
                    nums1(k) = nums1(i)
                    i -= 1
                Else
                    nums1(k) = nums2(j)
                    j -= 1
                End If
                k -= 1
            End While
            
            ' 如果nums2还有剩余元素，全部复制到nums1
            While j >= 0
                nums1(k) = nums2(j)
                j -= 1
                k -= 1
            End While
        End Sub
        
        ' 面试题5: 反转链表
        Public Shared Function ReverseLinkedList(head As ListNode) As ListNode
            Dim prev As ListNode = Nothing
            Dim current As ListNode = head
            
            While current IsNot Nothing
                Dim nextTemp As ListNode = current.Next ' 保存下一个节点
                current.Next = prev ' 反转指针
                prev = current ' 移动prev指针
                current = nextTemp ' 移动current指针
            End While
            
            Return prev ' 新的头结点
        End Function
        
        ' 面试题6: 使用栈实现队列
        Public Class MyQueue
            Private inStack As New Stack(Of Integer)() ' 用于入队操作
            Private outStack As New Stack(Of Integer)() ' 用于出队操作
            
            ' 入队操作
            Public Sub Push(x As Integer)
                inStack.Push(x)
            End Sub
            
            ' 出队操作
            Public Function Pop() As Integer
                If outStack.Count = 0 Then
                    ' 如果outStack为空，将inStack中的所有元素倒入outStack
                    While inStack.Count > 0
                        outStack.Push(inStack.Pop())
                    End While
                End If
                
                Return outStack.Pop()
            End Function
            
            ' 查看队首元素
            Public Function Peek() As Integer
                If outStack.Count = 0 Then
                    ' 如果outStack为空，将inStack中的所有元素倒入outStack
                    While inStack.Count > 0
                        outStack.Push(inStack.Pop())
                    End While
                End If
                
                Return outStack.Peek()
            End Function
            
            ' 检查队列是否为空
            Public Function Empty() As Boolean
                Return inStack.Count = 0 AndAlso outStack.Count = 0
            End Function
        End Class
        
        ' 面试题7: 二叉树的最大深度 (递归解法)
        Public Shared Function MaxDepth(root As TreeNode) As Integer
            If root Is Nothing Then
                Return 0
            End If
            
            Dim leftDepth As Integer = MaxDepth(root.Left)
            Dim rightDepth As Integer = MaxDepth(root.Right)
            
            ' 当前节点的最大深度 = max(左子树深度, 右子树深度) + 1
            Return Math.Max(leftDepth, rightDepth) + 1
        End Function
        
        ' 面试题8: 有效的括号
        Public Shared Function IsValidParentheses(s As String) As Boolean
            If String.IsNullOrEmpty(s) Then
                Return True
            End If
            
            Dim stack As New Stack(Of Char)()
            
            For Each c As Char In s
                If c = "("c OrElse c = "["c OrElse c = "{"c Then
                    stack.Push(c)
                Else
                    If stack.Count = 0 Then
                        Return False
                    End If
                    
                    Dim top As Char = stack.Pop()
                    
                    If c = ")"c AndAlso top <> "("c Then
                        Return False
                    End If
                    If c = "]"c AndAlso top <> "["c Then
                        Return False
                    End If
                    If c = "}"c AndAlso top <> "{"c Then
                        Return False
                    End If
                End If
            Next
            
            Return stack.Count = 0
        End Function
    End Class

    ' 支持类：链表节点
    Public Class ListNode
        Public Property Val As Integer
        Public Property [Next] As ListNode
        
        Public Sub New(Optional val As Integer = 0, Optional [next] As ListNode = Nothing)
            Me.Val = val
            Me.Next = [next]
        End Sub
    End Class
    
    ' 支持类：二叉树节点
    Public Class TreeNode
        Public Property Val As Integer
        Public Property Left As TreeNode
        Public Property Right As TreeNode
        
        Public Sub New(Optional val As Integer = 0, Optional left As TreeNode = Nothing, Optional right As TreeNode = Nothing)
            Me.Val = val
            Me.Left = left
            Me.Right = right
        End Sub
    End Class
End Namespace 